analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/BoolConstraint.cs

```diff
@@ -22,21 +22,26 @@ namespace SonarAnalyzer.SymbolicExecution.Constraints
 {
     public sealed class BoolConstraint : SymbolicConstraint
     {
-        public static readonly BoolConstraint True = new();
-        public static readonly BoolConstraint False = new();
+        public static readonly BoolConstraint True = new(ConstraintKind.BoolTrue);
+        public static readonly BoolConstraint False = new(ConstraintKind.BoolFalse);
 
         public override SymbolicConstraint Opposite =>
             this == True ? False : True;
 
-        protected override string Name =>
-            this == True ? nameof(True) : nameof(False);
-
-        private BoolConstraint() { }
+        private BoolConstraint(ConstraintKind kind) : base(kind) { }
 
         public static BoolConstraint From(bool value) =>
             value ? True : False;
 
         public bool Equals(bool value) =>
             this == From(value);
+
+        public override string ToString() =>
+            this switch
+            {
+                _ when ReferenceEquals(this, True) => nameof(True),
+                _ when ReferenceEquals(this, False) => nameof(False),
+                _ => base.ToString(),
```

Comment:
Let's be explicit that this should have never happened
```suggestion
                _ => throw new InvalidOperationException("Unexpected instance")
```

Reply:
This violates https://rules.sonarsource.com/csharp/RSPEC-3877
> It is expected that some methods should be called with caution, but others, such as ToString, are expected to "just work". Throwing an exception from such a method is likely to break callers' code unexpectedly.

I kept the base call instead.